subsys/net/ip/net_context.c

```diff
@@ -2281,6 +2222,125 @@ static int context_sendto(struct net_context *context,
 	return ret;
 }
 
+struct net_pkt *net_context_alloc_pkt(struct net_context *context, sa_family_t family,
+				      size_t *size, int *result)
+{
+	size_t tmp_size;
+	struct net_pkt *pkt;
+
+	if (!context || !size || !result || !PART_OF_ARRAY(contexts, context)) {
+		*result = -EINVAL;
+		return NULL;
+	}
+
+	pkt = context_alloc_pkt(context, family, *size, PKT_WAIT_TIME);
+
+	if (!pkt) {
+		NET_ERR("Failed to allocate net_pkt");
+		*result = -ENOBUFS;
+		return NULL;
+	}
+
+	tmp_size = net_pkt_available_payload_buffer(pkt, net_context_get_proto(context));
+	if (tmp_size < *size) {
+		if (net_context_get_type(context) == SOCK_DGRAM) {
+			NET_ERR(
+				"Available payload buffer (%zu) is not enough for requested DGRAM (%zu)",
+				tmp_size, *size);
+			if (pkt != NULL) {
+				net_pkt_unref(pkt);
+			}
+			*result = -ENOMEM;
+			return NULL;
+		}
+		*size = tmp_size;
+	}
+	*result = 0;
+
+	return pkt;
+}
+
+int net_context_write_data(struct net_pkt *pkt,
+			   const void *data, int length)
+{
+	if (!pkt || !data || !PART_OF_ARRAY(contexts, net_pkt_context(pkt))) {
+		return -EINVAL;
+	}
+
+	return net_pkt_write(pkt, data, length);
+}
+
+int net_context_setup_udp_packet(struct net_context *context,
+				 sa_family_t family,
+				 struct net_pkt *pkt,
+				 const struct sockaddr *dst_addr)
+{
+	int ret = -EINVAL;
+	uint16_t dst_port = 0U;
+
+	if (!pkt || !context || !dst_addr ||
+	    !PART_OF_ARRAY(contexts, net_pkt_context(pkt))) {
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		struct sockaddr_in6 *addr6 = (struct sockaddr_in6 *)dst_addr;
+
+		dst_port = addr6->sin6_port;
+
+		ret = net_context_create_ipv6_new(context, pkt,
+						  NULL, &addr6->sin6_addr);
+	} else if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+		struct sockaddr_in *addr4 = (struct sockaddr_in *)dst_addr;
+
+		dst_port = addr4->sin_port;
+
+		ret = net_context_create_ipv4_new(context, pkt,
+						  NULL, &addr4->sin_addr);
+	}
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = bind_default(context);
+	if (ret) {
+		return ret;
+	}
+
+	ret = net_udp_create(pkt,
+			     net_sin((struct sockaddr *)
+				     &context->local)->sin_port,
+			     dst_port);
+
+	return ret;
+}
+
+int net_context_finalize_packet(struct net_context *context,
```

Comment:
This function has its usage within `net_context.c`. However, I think that there is no gain for an user if one can use `net_ipvX_finalize()` functions directly. We can set a requirement for sending function to expect valid packet.

Reply:
This function is exist anyway as static, so it can be reused by user and use internally as well. I want to keep it.